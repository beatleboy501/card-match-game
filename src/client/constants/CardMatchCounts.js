export default Object.freeze({
  THREE_CARD_MATCH: 3,
  TWO_CARD_MATCH: 2,
});
