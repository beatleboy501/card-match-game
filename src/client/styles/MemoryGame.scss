.container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  flex-direction: column;
}

.cardsContainer {
  margin: 0 auto;
  max-width: 1200px;
  padding: 0 1rem;
}

.cards {
  justify-content: flex-start;
  align-content: stretch;
  padding: 0;
  max-width: 100%;
}

.matches {
  margin-top: 1em !important;
}
