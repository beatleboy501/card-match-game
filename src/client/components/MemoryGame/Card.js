import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import Flipper from './Flipper';
import sassStyles from '../../styles/Card.scss';
import logo from '../../../public/timesLogo.png';

const propTypes = {
  cardInfo: PropTypes.object.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired,
};

const defaultProps = {};

const styles = {
  card: {
    width: 100,
    height: 175,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  bullet: {
    display: 'inline-block',
    margin: '0 2px',
    transform: 'scale(0.8)',
  },
  title: {
    marginBottom: 16,
    fontSize: 14,
  },
  pos: {
    marginBottom: 12,
  },
  cardBack: {
    height: 50,
    width: 50,
  },
};

export const GameCard = (props) => {
  const { cardInfo, isOpen, onClick } = props;
  return (
    <div className={sassStyles.responsiveCell}>
      <Flipper isClosed={!isOpen}>
        <Card onClick={onClick} style={styles.card} key="front">
          <CardContent>
            <div>
              <Typography color="textSecondary">
                {' '}
              </Typography>
              <Typography
                variant="headline"
                component="h2"
              >
                {cardInfo.value}
              </Typography>
            </div>
          </CardContent>
        </Card>
        <Card onClick={onClick} style={styles.card} key="back">
          <CardContent>
            <CardMedia style={styles.cardBack} image={logo} />
          </CardContent>
        </Card>
      </Flipper>
    </div>
  );
};

GameCard.propTypes = propTypes;
GameCard.defaultProps = defaultProps;

export default withStyles(styles)(GameCard);
