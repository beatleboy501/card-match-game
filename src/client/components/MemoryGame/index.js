import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Typography from '@material-ui/core/Typography';
import AwesomeDebouncePromise from 'awesome-debounce-promise';
import Card from './Card';
import styles from '../../styles/MemoryGame.scss';
import {
  addMatch, addOpenId, setOpenIds, setMatchedCards, startGame,
} from '../../redux/actions/memoryGameActions';
import CardMatchCounts from '../../constants/CardMatchCounts';

const propTypes = {
  cards: PropTypes.arrayOf(PropTypes.object),
  isFirstCard: PropTypes.bool,
  dispatchStartGame: PropTypes.func.isRequired,
  dispatchAddOpenId: PropTypes.func.isRequired,
  dispatchAddMatch: PropTypes.func.isRequired,
  dispatchSetOpenIds: PropTypes.func.isRequired,
  onGameStart: PropTypes.func.isRequired,
  onGameComplete: PropTypes.func.isRequired,
  openIds: PropTypes.arrayOf(PropTypes.string),
  matchedCards: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.string)),
  cardMatchCount: PropTypes.number.isRequired,
};

const defaultProps = {
  cards: [],
  openIds: [],
  matchedCards: [],
  isFirstCard: true,
};

export class MemoryGame extends React.Component {
  constructor(props) {
    super(props);
    this.handleCardClick = AwesomeDebouncePromise(this.handleCardClick.bind(this), 500);
    this.handleCardClickResult = this.handleCardClickResult.bind(this);
    this.isMatched = this.isMatched.bind(this);
    this.isGameComplete = this.isGameComplete.bind(this);
    this.completeMatch = this.completeMatch.bind(this);
  }

  isGameComplete() {
    const { cards, matchedCards, cardMatchCount } = this.props;
    return (matchedCards.length && cards.length)
      && matchedCards.length >= cards.length / cardMatchCount;
  }

  isMatched(cardId) {
    const { matchedCards } = this.props;
    const matches = matchedCards.reduce((acc, match) => acc.concat(match), []);
    return matches.includes(cardId);
  }

  handleCardClick(clickedId) {
    const {
      openIds, isFirstCard, onGameStart, dispatchStartGame, cardMatchCount, dispatchAddOpenId,
    } = this.props;
    if (isFirstCard) {
      onGameStart();
      dispatchStartGame();
    }
    if (this.isMatched(clickedId)
      || openIds.includes(clickedId)
      || openIds.length >= cardMatchCount) return;
    if (openIds.length > 0) { // One card has already been flipped
      dispatchAddOpenId(clickedId);
      setTimeout(() => {
        this.handleCardClickResult(clickedId);
      }, 1000);
    } else { // No cards have been flipped
      dispatchAddOpenId(clickedId);
    }
  }

  handleCardClickResult(clickedId) {
    const {
      cards, dispatchSetOpenIds, openIds, matchedCards, cardMatchCount,
    } = this.props;
    if (cardMatchCount === CardMatchCounts.TWO_CARD_MATCH) {
      const clickedCard = cards.find(card => card.id === clickedId);
      const openId = openIds.find(id => id !== clickedId);
      const openCard = cards.find(card => card.id === openId);
      if ((openCard && clickedCard) && openCard.value === clickedCard.value) { // there's a match
        this.completeMatch([clickedId, openId], matchedCards);
      } else { // no match
        dispatchSetOpenIds([]);
      }
    } else if (cardMatchCount === CardMatchCounts.THREE_CARD_MATCH) {
      const openCards = cards.filter(card => openIds.includes(card.id));
      const openValues = openCards.map(card => card.value);
      const distinctValues = openValues.reduce((acc, value) => {
        if (!acc.includes(value)) acc.push(value);
        return acc;
      }, []);
      if (distinctValues.length === 1) { // there's a match
        if (openValues.length === cardMatchCount) {
          this.completeMatch(openIds, matchedCards);
        }
      } else { // no match
        dispatchSetOpenIds([]);
      }
    }
  }

  completeMatch(matchArray, matchedCards) {
    const { dispatchAddMatch, dispatchSetOpenIds, onGameComplete } = this.props;
    dispatchAddMatch(matchArray);
    dispatchSetOpenIds([]);
    if (this.isGameComplete()) {
      onGameComplete(matchedCards.length);
    }
  }

  render() {
    const { cards, matchedCards, openIds } = this.props;
    return (
      <div className={styles.container}>
        <Typography className={styles.matches} variant="title" color="inherit">
          Matches
          {' '}
          {' '}
:
          {' '}
          {matchedCards.length}
        </Typography>
        <div className={styles.cardsContainer}>
          <div className={styles.cards}>
            {cards.map(card => (
              <Card
                key={card.id}
                cardInfo={card}
                isOpen={// is part of a match or it has been clicked
                  this.isMatched(card.id) || openIds.includes(card.id)
                }
                onClick={() => this.handleCardClick(card.id)}
              />
            ))}
          </div>
        </div>
      </div>
    );
  }
}

MemoryGame.propTypes = propTypes;
MemoryGame.defaultProps = defaultProps;

function mapStateToProps(state) {
  return { ...state, ...state.memoryGame };
}

const mapDispatchToProps = dispatch => ({
  dispatchAddMatch: (match) => {
    dispatch(addMatch(match));
  },
  dispatchAddOpenId: (openId) => {
    dispatch(addOpenId(openId));
  },
  dispatchSetOpenIds: (openIds) => {
    dispatch(setOpenIds(openIds));
  },
  dispatchSetMatchedCards: (matchedCards) => {
    dispatch(setMatchedCards(matchedCards));
  },
  dispatchStartGame: () => {
    dispatch(startGame());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(MemoryGame);
