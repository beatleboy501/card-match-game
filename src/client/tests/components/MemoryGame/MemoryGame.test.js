import React from 'react';
import { Provider } from 'react-redux';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import { MemoryGame } from '../../../components/MemoryGame/index';
import { easyCards, resetProps } from '../../testUtils';
import CardMatchCounts from '../../../constants/CardMatchCounts';

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureStore();
const initialState = {
  cards: [],
  openIds: [],
  pairedCards: [],
  isFirstCard: true,
};

let store;
let component;
let defaultProps;

const mockGameStart = jest.fn(id => id);
const mockStartGame = jest.fn(() => {});
const mockGameComplete = jest.fn(() => {});
const mockAddOpenId = jest.fn(id => id);
const mockAddMatch = jest.fn(arr => arr);
const mockSetOpenIds = jest.fn(id => id);

beforeEach(() => {
  store = mockStore(initialState);
  defaultProps = {
    cards: easyCards,
    isFirstCard: true,
    cardMatchCount: CardMatchCounts.TWO_CARD_MATCH,
    dispatchStartGame: mockStartGame,
    dispatchAddOpenId: mockAddOpenId,
    dispatchAddMatch: mockAddMatch,
    dispatchSetOpenIds: mockSetOpenIds,
    onGameStart: mockGameStart,
    onGameComplete: mockGameComplete,
    openIds: [],
    matchedCards: [],
  };
});

afterEach(() => {
  defaultProps = resetProps(defaultProps);
});

it('Renders without crashing', () => {
  component = shallow(
    <Provider store={store}>
      <MemoryGame {...defaultProps} />
    </Provider>,
  ).shallow();
  expect(component).toMatchSnapshot();
});

it('does not mark the game as complete when insufficient number of cards are matched', () => {
  const props = {
    ...defaultProps,
    cards: easyCards,
    matchedCards: [['1', '2'], ['3', '4']],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  expect(component.instance().isGameComplete()).toBe(false);
});

it('marks the game as complete when correct number of matchedCards is reached', () => {
  const props = {
    ...defaultProps,
    cards: easyCards,
    matchedCards: [['1', '2'], ['3', '4'], ['5', '6'], ['7', '8']],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  expect(component.instance().isGameComplete()).toBe(true);
});

it('isMatched returns true if card is matched', () => {
  const props = {
    ...defaultProps,
    matchedCards: [['1', '2']],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  expect(component.instance().isMatched('2')).toBe(true);
});

it('isMatched returns false if card is not matched', () => {
  const props = {
    ...defaultProps,
    matchedCards: [['1', '2']],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  expect(component.instance().isMatched('3')).toBe(false);
});

it('onGameStart is fired when first card is clicked', async () => {
  component = mount(
    <MemoryGame {...defaultProps} />,
  );
  await component.instance().handleCardClick(1);
  expect(mockStartGame).toHaveBeenCalled();
  expect(mockGameStart).toHaveBeenCalled();
  expect(mockAddOpenId).toHaveBeenCalled();
});

it('onGameStart is not fired after first card has been clicked', async () => {
  const props = {
    ...defaultProps,
    isFirstCard: false,
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  await component.instance().handleCardClick(1);
  expect(mockStartGame).not.toHaveBeenCalled();
  expect(mockGameStart).not.toHaveBeenCalled();
  expect(mockAddOpenId).toHaveBeenCalled();
});

it('adds single card if two card match and no other cards open after one card has been clicked', async () => {
  const props = {
    ...defaultProps,
    isFirstCard: false,
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  await component.instance().handleCardClick(1);
  expect(mockAddOpenId).toHaveBeenCalledWith(1);
});

it('does nothing with click if card is already open', async () => {
  const props = {
    ...defaultProps,
    isFirstCard: false,
    openIds: [1],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  await component.instance().handleCardClick(1);
  expect(mockAddOpenId).not.toHaveBeenCalled();
});

it('does nothing if openIds contains more ids than the match count', async () => {
  const props = {
    ...defaultProps,
    isFirstCard: false,
    openIds: [1, 2, 3],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  await component.instance().handleCardClick(4);
  expect(mockSetOpenIds).not.toHaveBeenCalled();
});

it('resets if no match', async () => {
  const props = {
    ...defaultProps,
    isFirstCard: false,
    openIds: [1, 2],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  await component.instance().handleCardClickResult(2);
  expect(mockSetOpenIds).toHaveBeenCalledWith([]);
});

it('adds match if the open card matches the clicked card', async () => {
  const props = {
    ...defaultProps,
    cards: easyCards,
    isFirstCard: false,
    openIds: [1, 5],
  };
  component = mount(
    <MemoryGame {...props} />,
  );
  await component.instance().handleCardClickResult(5);
  expect(mockAddMatch).toHaveBeenCalledWith([5, 1]);
  expect(mockSetOpenIds).toHaveBeenCalledWith([]);
});
