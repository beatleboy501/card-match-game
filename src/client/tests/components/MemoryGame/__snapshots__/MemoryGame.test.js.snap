// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MemoryGame
    cardMatchCount={2}
    cards={
      Array [
        Object {
          "id": 1,
          "value": "a",
        },
        Object {
          "id": 2,
          "value": "b",
        },
        Object {
          "id": 3,
          "value": "c",
        },
        Object {
          "id": 4,
          "value": "d",
        },
        Object {
          "id": 5,
          "value": "a",
        },
        Object {
          "id": 6,
          "value": "b",
        },
        Object {
          "id": 7,
          "value": "c",
        },
        Object {
          "id": 8,
          "value": "d",
        },
      ]
    }
    dispatchAddMatch={[MockFunction]}
    dispatchAddOpenId={[MockFunction]}
    dispatchSetOpenIds={[MockFunction]}
    dispatchStartGame={[MockFunction]}
    isFirstCard={true}
    matchedCards={Array []}
    onGameComplete={[MockFunction]}
    onGameStart={[MockFunction]}
    openIds={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <WithStyles(Typography)
          className="matches"
          color="inherit"
          variant="title"
        >
          Matches
           
           
          :
           
          0
        </WithStyles(Typography)>,
        <div
          className="cardsContainer"
        >
          <div
            className="cards"
          >
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 1,
                  "value": "a",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 2,
                  "value": "b",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 3,
                  "value": "c",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 4,
                  "value": "d",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 5,
                  "value": "a",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 6,
                  "value": "b",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 7,
                  "value": "c",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 8,
                  "value": "d",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
          </div>
        </div>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "Matches",
            " ",
            " ",
            ":",
            " ",
            0,
          ],
          "className": "matches",
          "color": "inherit",
          "variant": "title",
        },
        "ref": null,
        "rendered": Array [
          "Matches",
          " ",
          " ",
          ":",
          " ",
          0,
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="cards"
          >
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 1,
                  "value": "a",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 2,
                  "value": "b",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 3,
                  "value": "c",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 4,
                  "value": "d",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 5,
                  "value": "a",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 6,
                  "value": "b",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 7,
                  "value": "c",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
            <WithStyles(GameCard)
              cardInfo={
                Object {
                  "id": 8,
                  "value": "d",
                }
              }
              isOpen={false}
              onClick={[Function]}
            />
          </div>,
          "className": "cardsContainer",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 1,
                    "value": "a",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 2,
                    "value": "b",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 3,
                    "value": "c",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 4,
                    "value": "d",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 5,
                    "value": "a",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 6,
                    "value": "b",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 7,
                    "value": "c",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 8,
                    "value": "d",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />,
            ],
            "className": "cards",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 1,
                  "value": "a",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 2,
                  "value": "b",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "3",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 3,
                  "value": "c",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "4",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 4,
                  "value": "d",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "5",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 5,
                  "value": "a",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "6",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 6,
                  "value": "b",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "7",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 7,
                  "value": "c",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "8",
              "nodeType": "class",
              "props": Object {
                "cardInfo": Object {
                  "id": 8,
                  "value": "d",
                },
                "isOpen": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            className="matches"
            color="inherit"
            variant="title"
          >
            Matches
             
             
            :
             
            0
          </WithStyles(Typography)>,
          <div
            className="cardsContainer"
          >
            <div
              className="cards"
            >
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 1,
                    "value": "a",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 2,
                    "value": "b",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 3,
                    "value": "c",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 4,
                    "value": "d",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 5,
                    "value": "a",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 6,
                    "value": "b",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 7,
                    "value": "c",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 8,
                    "value": "d",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
            </div>
          </div>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Matches",
              " ",
              " ",
              ":",
              " ",
              0,
            ],
            "className": "matches",
            "color": "inherit",
            "variant": "title",
          },
          "ref": null,
          "rendered": Array [
            "Matches",
            " ",
            " ",
            ":",
            " ",
            0,
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="cards"
            >
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 1,
                    "value": "a",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 2,
                    "value": "b",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 3,
                    "value": "c",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 4,
                    "value": "d",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 5,
                    "value": "a",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 6,
                    "value": "b",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 7,
                    "value": "c",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
              <WithStyles(GameCard)
                cardInfo={
                  Object {
                    "id": 8,
                    "value": "d",
                  }
                }
                isOpen={false}
                onClick={[Function]}
              />
            </div>,
            "className": "cardsContainer",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 1,
                      "value": "a",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 2,
                      "value": "b",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 3,
                      "value": "c",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 4,
                      "value": "d",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 5,
                      "value": "a",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 6,
                      "value": "b",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 7,
                      "value": "c",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
                <WithStyles(GameCard)
                  cardInfo={
                    Object {
                      "id": 8,
                      "value": "d",
                    }
                  }
                  isOpen={false}
                  onClick={[Function]}
                />,
              ],
              "className": "cards",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 1,
                    "value": "a",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "2",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 2,
                    "value": "b",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "3",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 3,
                    "value": "c",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "4",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 4,
                    "value": "d",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "5",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 5,
                    "value": "a",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "6",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 6,
                    "value": "b",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "7",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 7,
                    "value": "c",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "8",
                "nodeType": "class",
                "props": Object {
                  "cardInfo": Object {
                    "id": 8,
                    "value": "d",
                  },
                  "isOpen": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
